components/eamxx/src/diagnostics/tests/horiz_avg_test.cpp

```diff
@@ -0,0 +1,226 @@
+
+  ekat::ParameterList params;
+  // REQUIRE_THROWS(diag_factory.create("HorizAvgDiag", comm,
+  //                                    params));  // No 'field_name' parameter
+
+  // Set time for qc and randomize its values
+  qc1.get_header().get_tracking().update_time_stamp(t0);
+  qc2.get_header().get_tracking().update_time_stamp(t0);
+  qc3.get_header().get_tracking().update_time_stamp(t0);
+  qc4.get_header().get_tracking().update_time_stamp(t0);
+  randomize(qc1, engine, pdf);
+  randomize(qc2, engine, pdf);
+  randomize(qc3, engine, pdf);
+  randomize(qc4, engine, pdf);
+
+  // Create and set up the diagnostic
+  params.set("grid_name", grid1->name());
+  params.set<std::string>("field_name", "qc");
+  auto diag1 = diag_factory.create("HorizAvgDiag", comm, params);
+  auto diag2 = diag_factory.create("HorizAvgDiag", comm, params);
+  auto diag3 = diag_factory.create("HorizAvgDiag", comm, params);
+  auto diag4 = diag_factory.create("HorizAvgDiag", comm, params);
+  diag1->set_grids(gm1);
+  diag2->set_grids(gm2);
+  diag3->set_grids(gm2);
+  diag4->set_grids(gm2);
+
+  auto area = grid1->get_geometry_data("area");
+
+  diag1->set_required_field(qc1);
+  diag1->initialize(t0, RunType::Initial);
+
+  diag1->compute_diagnostic();
+  auto diag1_f = diag1->get_diagnostic();
+
+  FieldIdentifier diag0_fid("qc_horiz_avg_manual",
+                            scalar1d_layout.clone().strip_dim(COL), kg / kg,
+                            grid1->name());
+  Field diag0(diag0_fid);
+  diag0.allocate_view();
+  auto diag0_v = diag0.get_view<Real>();
+
+  auto qc1_v  = qc1.get_view<Real *>();
+  auto area_v = area.get_view<const Real *>();
+
+  // calculate total area
+  Real atot = 0.0;
+  Kokkos::parallel_reduce(
+      "HorizAvgDiag::compute_diagnostic_impl::total_area", ngcols,
+      KOKKOS_LAMBDA(const int icol, Real &local_atot) {
+        local_atot += area_v[icol];
+      },
+      atot);
+  // calculate weighted avg
+  Real wavg = 0.0;
+  Kokkos::parallel_reduce(
+      "HorizAvgDiag::compute_diagnostic_impl::weighted_sum", ngcols,
+      KOKKOS_LAMBDA(const int icol, Real &local_wavg) {
+        local_wavg += (area_v[icol] / atot) * qc1_v[icol];
+      },
+      wavg);
+  Kokkos::deep_copy(diag0_v, wavg);
+
+  auto diag1_v = diag1_f.get_view<Real>();
+
+  bool result = false;
+  Kokkos::parallel_reduce(
+      Kokkos::RangePolicy<>(0, 1),
+      KOKKOS_LAMBDA(const int, bool &local_result) {
+        local_result = (diag1_v() == diag0_v());
+      },
+      result);
+  REQUIRE(result);
+
+  // Try known cases
+  // Set qc1_v to 1.0 to get weighted average of 1.0
+  Kokkos::deep_copy(qc1_v, 1.0);
+  Kokkos::deep_copy(diag0_v, 1.0);
+  diag1->compute_diagnostic();
+  auto diag1_v2 = diag1_f.get_view<Real>();
+  Kokkos::parallel_reduce(
+      Kokkos::RangePolicy<>(0, 1),
+      KOKKOS_LAMBDA(const int, bool &local_result) {
+        local_result = (diag1_v2() == diag0_v());
+      },
+      result);
+  REQUIRE(result);
```

Comment:
Ah, this may be a case of small roundoff. You can double check with this:

```c++
#include <iomanip>
...
auto v1h = ...;
auto v2h = ...;
std::cout << "v1:";
for (size_t i=0; i<v1h.size(); ++i) {
  std::cout << " " << std::setprecision(17) << v1h[i];
} std::cout << "\n";
std::cout << "v1:";
for (size_t i=0; i<v2h.size(); ++i) {
  std::cout << " " << std::setprecision(17) << v2h[i];
} std::cout << "\n";
```
I bet you'll see some ulp's diff. If this is indeed a roundoff issue, you can compare with the [fp number comparison macros](https://github.com/catchorg/Catch2/blob/devel/docs/comparing-floating-point-numbers.md) that catch2 offers.

Reply:
If you want to stick with the exact numbers check, you must tweak the random inputs in such a way that the result is an integer. E.g., make it so that tot-area is a (not too large) pow of 2 (so that `a / atot` in the diag is exactly represented), and the inputs are also integers.